Elastic exporter get broken pipe with big records

**Describe the bug**

It seems we can easily break the connection to our elastic cluster if we have bigger messages.
I wanted to replay a scenario we observed during the Gameday on 10.07.2020, where elastic went oom and crash looped.
Furthermore Zeebe crashed due to big messages and went out of disk space.

In our benchmarks it seems it wents out of disk space because it is not able to export, since elastic connections are broken.

**To Reproduce**

Setup:

 * 3 partitions, 3 nodes, replication factor 3
 * load 100 wfi/s
 * variable size ~ 1 MB

It happens on different benchmarks that the exporter was not able to export the data, which means the broker went out of disk space.

**Expected behavior**

We can export and process the record without issues.

**Log/Stacktrace**


We see continuously the following messages:
```
Expand all | Collapse all{
 insertId: "sjoehgj2yh7gqhpu6"  
 jsonPayload: {
  @type: "type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent"   
  context: {…}   
  exception: "io.zeebe.exporter.ElasticsearchExporterException: Failed to flush bulk
	at io.zeebe.exporter.ElasticsearchClient.flush(ElasticsearchClient.java:151) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.exporter.ElasticsearchExporter.flush(ElasticsearchExporter.java:116) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.exporter.ElasticsearchExporter.flushAndReschedule(ElasticsearchExporter.java:103) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:118) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:204) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
Caused by: java.io.IOException: Broken pipe
	at org.elasticsearch.client.RestClient$SyncResponseListener.get(RestClient.java:964) ~[elasticsearch-rest-client-6.8.10.jar:6.8.10]
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:233) ~[elasticsearch-rest-client-6.8.10.jar:6.8.10]
	at io.zeebe.exporter.ElasticsearchClient.exportBulk(ElasticsearchClient.java:189) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.exporter.ElasticsearchClient.flush(ElasticsearchClient.java:148) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	... 8 more
Caused by: java.io.IOException: Broken pipe
	at sun.nio.ch.FileDispatcherImpl.write0(Native Method) ~[?:?]
	at sun.nio.ch.SocketDispatcher.write(Unknown Source) ~[?:?]
	at sun.nio.ch.IOUtil.writeFromNativeBuffer(Unknown Source) ~[?:?]
	at sun.nio.ch.IOUtil.write(Unknown Source) ~[?:?]
	at sun.nio.ch.IOUtil.write(Unknown Source) ~[?:?]
	at sun.nio.ch.SocketChannelImpl.write(Unknown Source) ~[?:?]
	at org.apache.http.impl.nio.codecs.AbstractContentEncoder.doWriteChunk(AbstractContentEncoder.java:171) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.codecs.AbstractContentEncoder.doWriteChunk(AbstractContentEncoder.java:164) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.codecs.AbstractContentEncoder.writeToChannel(AbstractContentEncoder.java:138) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.codecs.LengthDelimitedEncoder.write(LengthDelimitedEncoder.java:130) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.nio.entity.NStringEntity.produceContent(NStringEntity.java:175) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.nio.protocol.BasicAsyncRequestProducer.produceContent(BasicAsyncRequestProducer.java:125) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.client.MainClientExec.produceContent(MainClientExec.java:262) ~[httpasyncclient-4.1.4.jar:4.1.4]
	at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.produceContent(DefaultClientExchangeHandlerImpl.java:140) ~[httpasyncclient-4.1.4.jar:4.1.4]
	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.outputReady(HttpAsyncRequestExecutor.java:249) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.DefaultNHttpClientConnection.produceOutput(DefaultNHttpClientConnection.java:290) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.client.InternalIODispatch.onOutputReady(InternalIODispatch.java:86) ~[httpasyncclient-4.1.4.jar:4.1.4]
	at org.apache.http.impl.nio.client.InternalIODispatch.onOutputReady(InternalIODispatch.java:39) ~[httpasyncclient-4.1.4.jar:4.1.4]
	at org.apache.http.impl.nio.reactor.AbstractIODispatch.outputReady(AbstractIODispatch.java:145) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.BaseIOReactor.writable(BaseIOReactor.java:187) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:341) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:591) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
```

https://console.cloud.google.com/logs/viewer?interval=NO_LIMIT&authuser=1&project=zeebe-io&minLogLevel=0&expandAll=false&timestamp=2020-07-13T12:58:29.015000000Z&customFacets=&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.namespace_name%3D%22zell-test%22%0Alabels.k8s-pod%2Fapp%3D%22zell-test-zeebe%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Finstance%3D%22zell-test%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fname%3D%22zeebe-cluster%22&scrollTimestamp=2020-07-13T12:58:16.058549000Z&pinnedLogId=wmfgkvw4gzmjced5f&pinnedLogTimestamp=2020-07-13T12:57:50.182212Z

or different benchmark

https://console.cloud.google.com/logs/viewer?interval=NO_LIMIT&authuser=1&project=zeebe-io&minLogLevel=0&expandAll=false&timestamp=2020-07-13T13:20:56.077000000Z&customFacets=&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.namespace_name%3D%22zell-chaos-large%22%0Alabels.k8s-pod%2Fapp%3D%22zell-chaos-large-zeebe%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Finstance%3D%22zell-chaos-large%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fname%3D%22zeebe-cluster%22&scrollTimestamp=2020-07-13T13:20:45.942604000Z&pinnedLogId=sjoehgj2yh7gqhpu6&pinnedLogTimestamp=2020-07-13T13:20:44.402767Z

<details><summary>Full Stacktrace</summary>
 <p>

```
E 2020-07-13T13:20:31.688372Z Error on exporting record with key 2251799813702286 
E 2020-07-13T13:20:34.432168Z Unexpected exception occurred on periodically flushing bulk, will retry later. 
E 2020-07-13T13:20:39.380522Z Unexpected exception occurred on periodically flushing bulk, will retry later. 
E 2020-07-13T13:20:41.727210Z Error on exporting record with key 2251799813702286 
E 2020-07-13T13:20:44.402767Z Unexpected exception occurred on periodically flushing bulk, will retry later. 
E 2020-07-13T13:20:50.039588Z Unexpected exception occurred on periodically flushing bulk, will retry later. 
E 2020-07-13T13:20:51.744128Z Error on exporting record with key 2251799813702286 
E 2020-07-13T13:20:54.443096Z Unexpected exception occurred on periodically flushing bulk, will retry later. 
E 2020-07-13T13:20:59.437443Z Unexpected exception occurred on periodically flushing bulk, will retry later. 
E 2020-07-13T13:21:02.193694Z Error on exporting record with key 2251799813702286 
E 2020-07-13T13:21:04.478565Z Unexpected exception occurred on periodically flushing bulk, will retry later. 
```

</p>
</details>


**Environment:**
 - OS: k8s
 - Zeebe Version: SNAPSHOT
 - Configuration: elastic

[i have seen this now with our normal benchmarks:](https://console.cloud.google.com/logs/viewer?interval=NO_LIMIT&authuser=1&project=zeebe-io&minLogLevel=0&expandAll=false&timestamp=2020-08-11T10%3A02%3A25.151000000Z&customFacets&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.namespace_name%3D%22zeebe-chaos%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.container_name%3D%22zeebe-cluster%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.pod_name%3D%22zeebe-chaos-zeebe-1%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Atimestamp%3D%222020-08-11T10%3A02%3A18.699405000Z%22%0AinsertId%3D%22iqme9mng62l9ro8xt%22&scrollTimestamp=2020-08-11T10%3A02%3A58.837601000Z)

```

2020-08-11 12:02:18.699 CEST
Error on exporting record with key 4503599627373489
 
Expand all | Collapse all{
 insertId: "iqme9mng62l9ro8xt"  
 jsonPayload: {
  @type: "type.googleapis.com/google.devtools.clouderrorreporting.v1beta1.ReportedErrorEvent"   
  context: {…}   
  exception: "io.zeebe.exporter.ElasticsearchExporterException: Failed to flush bulk
	at io.zeebe.exporter.ElasticsearchClient.flush(ElasticsearchClient.java:148) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.exporter.ElasticsearchExporter.flush(ElasticsearchExporter.java:116) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.exporter.ElasticsearchExporter.export(ElasticsearchExporter.java:84) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.broker.exporter.stream.ExporterDirector$RecordExporter.export(ExporterDirector.java:354) ~[zeebe-broker-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.retry.BackOffRetryStrategy.run(BackOffRetryStrategy.java:51) ~[zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:118) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:204) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
Caused by: java.io.IOException: Broken pipe
	at org.elasticsearch.client.RestClient$SyncResponseListener.get(RestClient.java:964) ~[elasticsearch-rest-client-6.8.11.jar:6.8.11]
	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:233) ~[elasticsearch-rest-client-6.8.11.jar:6.8.11]
	at io.zeebe.exporter.ElasticsearchClient.exportBulk(ElasticsearchClient.java:184) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.exporter.ElasticsearchClient.flush(ElasticsearchClient.java:145) ~[zeebe-elasticsearch-exporter-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	... 10 more
Caused by: java.io.IOException: Broken pipe
	at sun.nio.ch.FileDispatcherImpl.write0(Native Method) ~[?:?]
	at sun.nio.ch.SocketDispatcher.write(Unknown Source) ~[?:?]
	at sun.nio.ch.IOUtil.writeFromNativeBuffer(Unknown Source) ~[?:?]
	at sun.nio.ch.IOUtil.write(Unknown Source) ~[?:?]
	at sun.nio.ch.IOUtil.write(Unknown Source) ~[?:?]
	at sun.nio.ch.SocketChannelImpl.write(Unknown Source) ~[?:?]
	at org.apache.http.impl.nio.codecs.AbstractContentEncoder.doWriteChunk(AbstractContentEncoder.java:171) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.codecs.AbstractContentEncoder.doWriteChunk(AbstractContentEncoder.java:164) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.codecs.AbstractContentEncoder.writeToChannel(AbstractContentEncoder.java:138) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.codecs.LengthDelimitedEncoder.write(LengthDelimitedEncoder.java:130) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.nio.entity.NStringEntity.produceContent(NStringEntity.java:175) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.nio.protocol.BasicAsyncRequestProducer.produceContent(BasicAsyncRequestProducer.java:125) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.client.MainClientExec.produceContent(MainClientExec.java:262) ~[httpasyncclient-4.1.4.jar:4.1.4]
	at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.produceContent(DefaultClientExchangeHandlerImpl.java:140) ~[httpasyncclient-4.1.4.jar:4.1.4]
	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.outputReady(HttpAsyncRequestExecutor.java:249) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.DefaultNHttpClientConnection.produceOutput(DefaultNHttpClientConnection.java:290) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.client.InternalIODispatch.onOutputReady(InternalIODispatch.java:86) ~[httpasyncclient-4.1.4.jar:4.1.4]
	at org.apache.http.impl.nio.client.InternalIODispatch.onOutputReady(InternalIODispatch.java:39) ~[httpasyncclient-4.1.4.jar:4.1.4]
	at org.apache.http.impl.nio.reactor.AbstractIODispatch.outputReady(AbstractIODispatch.java:145) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.BaseIOReactor.writable(BaseIOReactor.java:187) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:341) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:591) ~[httpcore-nio-4.4.13.jar:4.4.13]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
```

--- a/broker/src/test/resources/system/elasticexporter.yaml
+++ b/broker/src/test/resources/system/elasticexporter.yaml
@@ -1,4 +1,3 @@
-
 zeebe:
   broker:
     exporters:
@@ -16,6 +15,7 @@ zeebe:
         #   bulk:
         #     delay: 5
         #     size: 1000
+        #     memoryLimit: 10485760
         #
           authentication:
             username: elastic

--- a/exporters/elasticsearch-exporter/README.md
+++ b/exporters/elasticsearch-exporter/README.md
@@ -38,24 +38,27 @@ it should be flushed (regardless of size) can be controlled by configuration.
 For example:
 
 ```yaml
-...  
+...
   exporters:
     elasticsearch:
       args:
         delay: 5
         size: 1000
+        memoryLimit: 10485760
 ```
 
 With the above example, the exporter would aggregate records and flush them to Elasticsearch
 either:
   1. when it has aggregated 1000 records
-  2. 5 seconds have elapsed since the last flush (regardless of how many records were aggregated)
+  2. when the batch memory size exceeds 10 MB
+  3. 5 seconds have elapsed since the last flush (regardless of how many records were aggregated)
 
 More specifically, each option configures the following:
 
 * `delay` (`integer`): a specific delay, in seconds, before we force flush the current batch. This ensures
 that even when we have low traffic of records we still export every once in a while.
-* `size` (`integer`): how big a batch should be before we export.
+* `size` (`integer`): how many records a batch should have before we export.
+* `memoryLimit` (`integer`): the size of the bulk, in bytes, before we export.
 
 ### Index
 
@@ -74,11 +77,11 @@ For example:
         index:
           prefix: zeebe-record
           createTemplate: true
-          
+
           command: false
           event: true
           rejection: false
-          
+
           deployment: false
           incident: true
           job: false
@@ -118,28 +121,29 @@ Here is a complete, default configuration example:
       #
       # These setting can also be overridden using the environment variables "ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_..."
       #
-      
+
       className: io.zeebe.exporter.ElasticsearchExporter
-     
+
       args:
         url: http://localhost:9200
-     
+
         bulk:
           delay: 5
           size: 1000
-     
+          memoryLimit: 10485760
+
         authentication:
           username: elastic
           password: changeme
-     
+
         index:
           prefix: zeebe-record
           createTemplate: true
-     
+
           command: false
           event: true
           rejection: false
-     
+
           deployment: true
           error: true
           incident: true
@@ -152,7 +156,7 @@ Here is a complete, default configuration example:
           workflowInstance: true
           workflowInstanceCreation: false
           workflowInstanceSubscription: false
-     
+
           ignoreVariablesAbove: 32677
 
 ```

--- a/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchClient.java
+++ b/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchClient.java
@@ -140,6 +140,9 @@ public class ElasticsearchClient {
     final int bulkSize = bulkRequest.size();
     metrics.recordBulkSize(bulkSize);
 
+    final var bulkMemorySize = getBulkMemorySize();
+    metrics.recordBulkMemorySize(bulkMemorySize);
+
     final BulkResponse bulkResponse;
     try {
       bulkResponse = exportBulk();
@@ -188,7 +191,12 @@ public class ElasticsearchClient {
   }
 
   public boolean shouldFlush() {
-    return bulkRequest.size() >= configuration.bulk.size;
+    return bulkRequest.size() >= configuration.bulk.size
+        || getBulkMemorySize() >= configuration.bulk.memoryLimit;
+  }
+
+  private int getBulkMemorySize() {
+    return bulkRequest.stream().mapToInt(String::length).sum();
   }
 
   /** @return true if request was acknowledged */

--- a/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java
+++ b/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java
@@ -22,6 +22,9 @@ public class ElasticsearchExporter implements Exporter {
 
   public static final String ZEEBE_RECORD_TEMPLATE_JSON = "/zeebe-record-template.json";
 
+  // by default, the bulk request may not be bigger than 100MB
+  private static final int RECOMMENDED_MAX_BULK_MEMORY_LIMIT = 100 * 1024 * 1024;
+
   private Logger log;
   private Controller controller;
 
@@ -91,6 +94,12 @@ public class ElasticsearchExporter implements Exporter {
           String.format(
               "Elasticsearch prefix must not contain underscore. Current value: %s",
               configuration.index.prefix));
+    }
+
+    if (configuration.bulk.memoryLimit > RECOMMENDED_MAX_BULK_MEMORY_LIMIT) {
+      log.warn(
+          "The bulk memory limit is set to more than {} bytes. It is recommended to set the limit between 5 to 15 MB.",
+          RECOMMENDED_MAX_BULK_MEMORY_LIMIT);
     }
   }
 

--- a/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporterConfiguration.java
+++ b/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporterConfiguration.java
@@ -166,10 +166,19 @@ public class ElasticsearchExporterConfiguration {
     public int delay = 5;
     // bulk size before flush
     public int size = 1_000;
+    // memory limit of the bulk in bytes before flush
+    public int memoryLimit = 10 * 1024 * 1024;
 
     @Override
     public String toString() {
-      return "BulkConfiguration{" + "delay=" + delay + ", size=" + size + '}';
+      return "BulkConfiguration{"
+          + "delay="
+          + delay
+          + ", size="
+          + size
+          + ", memoryLimit="
+          + memoryLimit
+          + '}';
     }
   }
 

--- a/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchMetrics.java
+++ b/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchMetrics.java
@@ -7,6 +7,7 @@
  */
 package io.zeebe.exporter;
 
+import io.prometheus.client.Gauge;
 import io.prometheus.client.Histogram;
 
 public class ElasticsearchMetrics {
@@ -28,6 +29,14 @@ public class ElasticsearchMetrics {
           .labelNames("partition")
           .register();
 
+  private static final Gauge BULK_MEMORY_SIZE =
+      Gauge.build()
+          .namespace("zeebe_elasticsearch_exporter")
+          .name("bulk_memory_size")
+          .help("Exporter bulk memory size")
+          .labelNames("partition")
+          .register();
+
   private final String partitionIdLabel;
 
   public ElasticsearchMetrics(final int partitionId) {
@@ -40,5 +49,9 @@ public class ElasticsearchMetrics {
 
   public void recordBulkSize(final int bulkSize) {
     BULK_SIZE.labels(partitionIdLabel).observe(bulkSize);
+  }
+
+  public void recordBulkMemorySize(final int bulkMemorySize) {
+    BULK_MEMORY_SIZE.labels(partitionIdLabel).set(bulkMemorySize);
   }
 }

--- a/exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchClientTest.java
+++ b/exporters/elasticsearch-exporter/src/test/java/io/zeebe/exporter/ElasticsearchClientTest.java
@@ -21,6 +21,7 @@ import io.zeebe.protocol.record.ValueType;
 import io.zeebe.protocol.record.value.VariableRecordValue;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.function.Function;
 import java.util.stream.IntStream;
 import org.junit.Before;
 import org.junit.Test;
@@ -163,5 +164,44 @@ public class ElasticsearchClientTest extends AbstractElasticsearchExporterIntegr
 
     // then
     assertThat(bulkRequest).hasSize(1);
+  }
+
+  @Test
+  public void shouldFlushOnMemoryLimit() {
+    // given
+    final var bulkMemoryLimit = 1024;
+    final var recordSize = 2;
+
+    configuration.bulk.memoryLimit = bulkMemoryLimit;
+    configuration.bulk.size = Integer.MAX_VALUE;
+    configuration.bulk.delay = Integer.MAX_VALUE;
+
+    final var variableValue1 = "x".repeat(bulkMemoryLimit / recordSize);
+    final var variableValue2 = "y".repeat(bulkMemoryLimit / recordSize);
+    final Function<String, String> jsonRecord =
+        (String value) -> String.format("{\"value\":\"%s\"}", value);
+
+    final VariableRecordValue recordValue = mock(VariableRecordValue.class);
+    when(recordValue.getValue()).thenReturn(variableValue1);
+
+    final Record<VariableRecordValue> recordMock = mock(Record.class);
+    when(recordMock.getKey()).thenReturn(1L);
+    when(recordMock.getPartitionId()).thenReturn(1);
+    when(recordMock.getValueType()).thenReturn(ValueType.VARIABLE);
+    when(recordMock.getValue()).thenReturn(recordValue);
+    when(recordMock.toJson()).thenReturn(jsonRecord.apply(variableValue1));
+
+    // when
+    client.index(recordMock);
+
+    assertThat(client.shouldFlush()).isFalse();
+
+    when(recordMock.getKey()).thenReturn(2L);
+    when(recordMock.toJson()).thenReturn(jsonRecord.apply(variableValue2));
+
+    client.index(recordMock);
+
+    // then
+    assertThat(client.shouldFlush()).isTrue();
   }
 }

--- a/monitor/grafana/zeebe.json
+++ b/monitor/grafana/zeebe.json
@@ -77,7 +77,7 @@
   "gnetId": null,
   "graphTooltip": 1,
   "id": null,
-  "iteration": 1596196141101,
+  "iteration": 1597737695016,
   "links": [],
   "panels": [
     {
@@ -7187,10 +7187,10 @@
           "dataFormat": "tsbuckets",
           "datasource": "$DS_PROMETHEUS",
           "gridPos": {
-            "h": 9,
-            "w": 12,
+            "h": 10,
+            "w": 8,
             "x": 0,
-            "y": 39
+            "y": 51
           },
           "heatmap": {},
           "hideZeroBuckets": true,
@@ -7200,7 +7200,6 @@
             "show": false
           },
           "links": [],
-          "options": {},
           "reverseYBuckets": false,
           "targets": [
             {
@@ -7251,10 +7250,10 @@
           "dataFormat": "tsbuckets",
           "datasource": "$DS_PROMETHEUS",
           "gridPos": {
-            "h": 9,
-            "w": 12,
-            "x": 12,
-            "y": 39
+            "h": 10,
+            "w": 8,
+            "x": 8,
+            "y": 51
           },
           "heatmap": {},
           "hideZeroBuckets": true,
@@ -7264,7 +7263,6 @@
             "show": false
           },
           "links": [],
-          "options": {},
           "reverseYBuckets": false,
           "targets": [
             {
@@ -7306,6 +7304,101 @@
           "dashLength": 10,
           "dashes": false,
           "datasource": "$DS_PROMETHEUS",
+          "fill": 1,
+          "fillGradient": 0,
+          "gridPos": {
+            "h": 10,
+            "w": 8,
+            "x": 16,
+            "y": 51
+          },
+          "hiddenSeries": false,
+          "id": 185,
+          "legend": {
+            "alignAsTable": false,
+            "avg": false,
+            "current": false,
+            "max": false,
+            "min": false,
+            "rightSide": false,
+            "show": true,
+            "total": false,
+            "values": false
+          },
+          "lines": true,
+          "linewidth": 1,
+          "links": [],
+          "nullPointMode": "null",
+          "options": {
+            "dataLinks": []
+          },
+          "percentage": false,
+          "pointradius": 2,
+          "points": false,
+          "renderer": "flot",
+          "seriesOverrides": [],
+          "spaceLength": 10,
+          "stack": false,
+          "steppedLine": false,
+          "targets": [
+            {
+              "expr": "zeebe_elasticsearch_exporter_bulk_memory_size{namespace=~\"$namespace\",pod=~\"$pod\",partition=~\"$partition\"}",
+              "format": "time_series",
+              "interval": "",
+              "intervalFactor": 1,
+              "legendFormat": "{{pod}} p{{partition}}",
+              "refId": "A"
+            }
+          ],
+          "thresholds": [],
+          "timeFrom": null,
+          "timeRegions": [],
+          "timeShift": null,
+          "title": "Elasticsearch Exporter (Bulk Memory Size)",
+          "tooltip": {
+            "shared": true,
+            "sort": 0,
+            "value_type": "individual"
+          },
+          "type": "graph",
+          "xaxis": {
+            "buckets": null,
+            "mode": "time",
+            "name": null,
+            "show": true,
+            "values": []
+          },
+          "yaxes": [
+            {
+              "$$hashKey": "object:359",
+              "format": "bytes",
+              "label": null,
+              "logBase": 1,
+              "max": null,
+              "min": null,
+              "show": true
+            },
+            {
+              "$$hashKey": "object:360",
+              "format": "short",
+              "label": null,
+              "logBase": 1,
+              "max": null,
+              "min": null,
+              "show": true
+            }
+          ],
+          "yaxis": {
+            "align": false,
+            "alignLevel": null
+          }
+        },
+        {
+          "aliasColors": {},
+          "bars": false,
+          "dashLength": 10,
+          "dashes": false,
+          "datasource": "$DS_PROMETHEUS",
           "editable": true,
           "error": false,
           "fill": 1,
@@ -7315,9 +7408,10 @@
             "h": 10,
             "w": 24,
             "x": 0,
-            "y": 48
+            "y": 61
           },
           "height": "400",
+          "hiddenSeries": false,
           "id": 52,
           "legend": {
             "alignAsTable": true,
@@ -7843,5 +7937,5 @@
   "variables": {
     "list": []
   },
-  "version": 9
+  "version": 11
 }

